File: benchmarks/alpaca/src/JsonParser.scala
import alpaca.*

val JsonLexer = lexer {
  // ignoring whitespaces
  case "\\s+" => Token.Ignored

  // brackets and punctuation marks
  case x @ ("\\{" | "\\}" | "\\[" | "\\]" | ":" | ",") => Token[x.type]

  // literals
  case x @ ("false" | "true") => Token["Bool"](x.toBoolean)
  case "null"                 => Token["Null"](null)

  // numbers and strings
  case x @ """[-+]?\d+(\.\d+)?""" => Token["Number"](x.toDouble)
  case x @ """"(\\.|[^"])*""""    => Token["String"](x.slice(1, x.length - 1))
}

object JsonParser extends Parser {
  val root: Rule[Any] = rule { case Value(value) => value }

  val Value: Rule[Any] = rule(
    { case JsonLexer.Null(n) => n.value },
    { case JsonLexer.Bool(b) => b.value },
    { case JsonLexer.Number(n) => n.value },
    { case JsonLexer.String(s) => s.value },
    { case Object(obj) => obj },
    { case Array(arr) => arr }
  )

  val Object: Rule[Map[String, Any]] = rule(
    { case (JsonLexer.`\\{`(_), JsonLexer.`\\}`(_)) => Map.empty[String, Any] },
    { case (JsonLexer.`\\{`(_), ObjectMembers(members), JsonLexer.`\\}`(_)) =>
      members.toMap
    }
  )

  val ObjectMembers: Rule[List[(String, Any)]] = rule(
    { case ObjectMember(member) => scala.List(member) },
    { case (ObjectMembers(members), JsonLexer.`,`(_), ObjectMember(member)) =>
      members :+ member
    }
  )

  val ObjectMember: Rule[(String, Any)] = rule {
    case (JsonLexer.String(s), JsonLexer.`:`(_), Value(v)) =>
      (s.value, v)
  }

  val Array: Rule[List[Any]] = rule(
    { case (JsonLexer.`\\[`(_), JsonLexer.`\\]`(_)) => Nil },
    { case (JsonLexer.`\\[`(_), ArrayElements(elems), JsonLexer.`\\]`(_)) =>
      elems
    }
  )

  val ArrayElements: Rule[List[Any]] = rule(
    { case Value(v) => scala.List(v) },
    { case (ArrayElements(elems), JsonLexer.`,`(_), Value(v)) => elems :+ v }
  )
}

object JsonParserMain extends App {
  import java.nio.file.{Files, Paths}

  val filePathIterative = s"inputs/iterative_json_3.txt"
  val fileContentIterative = new String(
    Files.readAllBytes(Paths.get(filePathIterative))
  )

  try {
    val (_, tokens) = JsonLexer.tokenize(fileContentIterative)
    val (_, result) = JsonParser.parse(tokens)
    println(s"\nResult Iterative: $result")
  } catch {
    case e: Exception =>
      println(s"\nError Iterative: ${e.getMessage}")
      e.printStackTrace()
  }

  val filePathRecursive = s"inputs/recursive_json_3.txt"
  val fileContentRecursive = new String(
    Files.readAllBytes(Paths.get(filePathRecursive))
  )

  try {
    val (_, tokens) = JsonLexer.tokenize(fileContentRecursive)
    val (_, result) = JsonParser.parse(tokens)
    println(s"\nResult Recursive: $result")
  } catch {
    case e: Exception =>
      println(s"\nError Recursive: ${e.getMessage}")
      e.printStackTrace()
  }
}

File: benchmarks/sly/json_parser.py
from pathlib import Path
from sly import Lexer, Parser

class JsonLexer(Lexer):
    literals = ['{', '}', '[', ']', ':', ',']
    tokens = {NULL, BOOL, NUM, STRING}
    ignore = ' \t\n'

    NULL = r'null'
    BOOL = r'true|false'
    NUM = r'[-+]?\d+(\.\d+)?'
    STRING = r'\"(\\.|[^"\\])*\"'
    
    def error(self, t):
        print(f"Illegal character '{t.value[0]}'")
        self.index += 1


class JsonParser(Parser):
    tokens = JsonLexer.tokens

    @_('value')
    def program(self, p):
        return p.value
    
    @_('NULL')
    def value(self, p):
        return None
    
    @_('BOOL')
    def value(self, p):
        return p.BOOL == "true"
    
    @_('NUM')
    def value(self, p):
        return float(p.NUM)
    
    @_('STRING')
    def value(self, p):
        return p.STRING[1:-1]

    @_('object')
    def value(self, p):
        return p.object
    
    @_('array')
    def value(self, p):
        return p.array
    
    @_('"{" "}"')
    def object(self, p):
        return {}

    @_('"{" members "}"')
    def object(self, p):
        return dict(p.members)
    
    @_('member')
    def members(self, p):
        return [p.member]
    
    @_('members "," member')
    def members(self, p):
        return p.members + [p.member]
    
    @_('STRING ":" value')
    def member(self, p):
        return (p.STRING[1:-1], p.value)
    
    @_('"[" "]"')
    def array(self, p):
        return []
    
    @_('"[" elements "]"')
    def array(self, p):
        return p.elements
    
    @_('value')
    def elements(self, p):
        return [p.value]
    
    @_('elements "," value')
    def elements(self, p):
        return p.elements + [p.value]

    def error(self, p):
        if p:
            print(f"Syntax error at token {p.type} ('{p.value}')")
        else:
            print("Syntax error at EOF")


if __name__ == "__main__":
    # Test the parser
    lexer = JsonLexer()
    parser = JsonParser()

    file_path = Path(__file__).parent.parent / "inputs" / "iterative_json_3.txt"
    with file_path.open('r') as f:
        file_content = f.read()
    try:
        tokens = lexer.tokenize(file_content)
        result = parser.parse(tokens)
        print(f"Result Iterative: {result}")
    except Exception as e:
        print(f"Error Iterative: {e}")


    file_path = Path(__file__).parent.parent / "inputs" / "recursive_json_3.txt"
    with file_path.open('r') as f:
        file_content = f.read()
    try:
        tokens = lexer.tokenize(file_content)
        result = parser.parse(tokens)
        print(f"Result Recursive: {result}")
    except Exception as e:
        print(f"Error Recursive: {e}")
